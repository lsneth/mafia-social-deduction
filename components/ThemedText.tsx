import { Text, type TextProps } from 'react-native'

export type ThemedTextProps = TextProps & {
  type?: 'default' | 'title' | 'subtitle'
  underline?: boolean
}

export function ThemedText({ className, type = 'default', underline = false, ...rest }: ThemedTextProps): JSX.Element {
  return (
    <Text
      style={{ fontFamily: type === 'title' ? 'Oswald_700Bold' : 'CrimsonText_400Regular' }}
      className={`text-mafiaWhite ${
        type === 'title' ? 'text-7xl pt-5' : type === 'subtitle' ? 'text-3xl' : 'text-lg'
      } ${underline ? 'underline' : ''} ${className}`}
      {...rest}
    />
  )
}
