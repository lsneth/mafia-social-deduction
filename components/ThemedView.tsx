import useAudio from '@/hooks/useAudio'
import { useGame } from '@/providers/GameProvider'
import { AVPlaybackSource } from 'expo-av'
import { useEffect } from 'react'
import { ImageBackground, ImageSourcePropType, View, type ViewProps } from 'react-native'
import Animated, { useSharedValue, withTiming } from 'react-native-reanimated'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

type ThemedViewProps = ViewProps & {
  bgImageSrc?: ImageSourcePropType
  fadeIn?: boolean
  fadeInAudio?: AVPlaybackSource | null
  preFadeInAudio?: AVPlaybackSource | null
}

export default function ThemedView({
  bgImageSrc,
  className,
  fadeIn = false,
  preFadeInAudio = null,
  fadeInAudio = null,
  ...rest
}: ThemedViewProps): JSX.Element {
  const insets = useSafeAreaInsets()
  const opacity = useSharedValue(fadeIn ? 0 : 1)
  const playFadeInAudio = useAudio(fadeInAudio)
  const playPreFadeInAudio = useAudio(preFadeInAudio)
  const { game, player } = useGame()
  const isHost = game?.host_id === player?.profile_id

  function startFadeIn() {
    opacity.value = withTiming(1, { duration: 5000 })
  }

  useEffect(() => {
    if (preFadeInAudio) {
      isHost && playPreFadeInAudio() // we only want to play the audio on the host's device
      setTimeout(() => {
        fadeInAudio && isHost && playFadeInAudio()
        startFadeIn()
      }, 5000)
    } else {
      fadeInAudio && isHost && playFadeInAudio()
      startFadeIn()
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps -- we only ever want to run this once
  }, [])

  return bgImageSrc ? (
    <Animated.View className="flex-1 bg-mafiaBlack" style={{ opacity }}>
      <ImageBackground source={bgImageSrc} className="h-full w-full">
        <View
          style={{
            paddingTop: insets.top,
            paddingBottom: insets.bottom,
          }}
          className="flex-1"
        >
          <View className={`flex-1 items-center p-8 ${className}`} {...rest} />
        </View>
      </ImageBackground>
    </Animated.View>
  ) : (
    <Animated.View
      style={{
        paddingTop: insets.top,
        paddingBottom: insets.bottom,
        opacity,
      }}
      className="flex-1 bg-mafiaBlack"
    >
      <View className={`flex-1 items-center p-8 ${className}`} {...rest} />
    </Animated.View>
  )
}
