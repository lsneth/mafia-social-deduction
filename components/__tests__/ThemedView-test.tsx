import * as React from 'react'
import renderer from 'react-test-renderer'

import { ThemedView } from '../ThemedView'
import { ThemedText } from '../ThemedText'

it(`renders correctly`, () => {
  const tree = renderer
    .create(
      <ThemedView>
        <ThemedText>This is a unit test for the ThemedView component</ThemedText>
      </ThemedView>
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})
