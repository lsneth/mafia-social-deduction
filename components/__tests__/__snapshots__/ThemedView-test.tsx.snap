// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#000",
      },
    ]
  }
>
  <Text
    style={
      [
        {
          "color": "#fff",
        },
        {
          "fontSize": 18,
          "lineHeight": 28,
        },
        {
          "textAlign": "center",
        },
        {
          "fontFamily": "CrimsonText_400Regular",
        },
      ]
    }
  >
    This is a unit test for the ThemedView component
  </Text>
</View>
`;
